lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs

```diff
@@ -346,3 +391,84 @@ spec = do
                 }
             ]
         }
+
+newtype SignedTx = SignedTx (Tx, [TxWitness])
+    deriving (Eq, Show, Generic)
+
+instance Arbitrary SignedTx where
+    arbitrary = do
+        nIns <- fromIntegral <$> arbitrary @Word8
+        nOut <- fromIntegral <$> arbitrary @Word8
+        inps <- vectorOf nIns arbitrary
+        outs <- vectorOf nOut arbitrary
+        wits <- vectorOf nIns arbitrary
+        let tid = fragmentId inps outs wits
+        return $ SignedTx (Tx tid inps outs, wits)
+
+    shrink (SignedTx (Tx _ inps outs, wits)) =
+        [ SignedTx (Tx (fragmentId inps' outs wits') inps' outs, wits')
+        | (inps', wits') <- unzip <$> shrinkList' (zip inps wits)
+        ]
+        ++
+        [ SignedTx (Tx (fragmentId inps outs' wits) inps outs', wits)
+        | outs' <- shrinkList' outs
+        ]
+
+      where
+        shrinkList' xs  =
+            (shrinkHeadAndReplicate shrink xs) ++
+            (shrinkList shrink xs)
+
+        -- Try shrinking the 'head' of the list and replace the elements in
+        -- the 'tail' with the exact same element. If the failure is related
+        -- to the size of the list, this makes the shrinking much faster.
+        shrinkHeadAndReplicate f (x:xs) =
+            (\x' -> x':(map (const x') xs)) <$> f x
+        shrinkHeadAndReplicate _f [] = []
+
+-- | Only generates single address witnesses
+instance Arbitrary TxWitness where
+    arbitrary = TxWitness <$> genFixed 64
+    shrink (TxWitness bytes) = TxWitness <$> shrinkFixedBS bytes
+
+instance Arbitrary (Hash "Tx") where
+    arbitrary = Hash <$> genFixed 32
+    shrink (Hash bytes) = Hash <$> shrinkFixedBS bytes
+
+instance Arbitrary Coin where
+    arbitrary = do
+        n <- choose (0, 100)
+        oneof [
+            return $ Coin n
+            , return $ Coin (getCoin (maxBound :: Coin) - n)
+            , Coin <$> choose (0, getCoin (maxBound :: Coin))
+            ]
+    shrink (Coin c) = map Coin (shrink c)
+
+instance Arbitrary TxIn where
+    arbitrary = TxIn
+        <$> arbitrary
+        <*> (fromIntegral <$> arbitrary @Word8)
+    shrink = genericShrink
+
+instance Arbitrary TxOut where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+-- Only generating single addresses!
+instance Arbitrary Address where
+    arbitrary = Address . prependTag 3 <$> genFixed 32
```

Comment:
Where does the value `3` come from?

Reply:
so the answer is in https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs#L509
I am generating here single addresses that have `3` header, hence prepending with this. Moreover, I am prepending to randomly generated `32` bytes as we have https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs#L529
Hope it answers your question

Reply:
added to the code the following in order to avoid magic number : 
```
466         let singleAddressHeader = 3                                                                                                                                                       
467         let singleAddressLenWithoutHeader = 32
```